core/trino-main/src/main/java/io/trino/operator/output/BlockBuilderPositionsAppender.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.BlockBuilderStatus;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.type.Type;
+import io.trino.type.BlockTypeOperators.BlockPositionEqual;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class BlockBuilderPositionsAppender
+        implements BlockTypeAwarePositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BlockBuilderPositionsAppender.class).instanceSize();
+
+    protected final Type type;
+    protected final BlockPositionEqual equalOperator;
+    private final BlockBuilder blockBuilder;
+
+    BlockBuilderPositionsAppender(Type type, BlockPositionEqual equalOperator, BlockBuilderStatus blockBuilderStatus, int expectedPositions)
+    {
+        this.type = requireNonNull(type, "type is null");
+        this.blockBuilder = type.createBlockBuilder(blockBuilderStatus, expectedPositions);
+        this.equalOperator = requireNonNull(equalOperator, "equalOperator is null");
+    }
+
+    public static class TypeBlockBuilderPositionsAppender
+            extends BlockBuilderPositionsAppender
+    {
+        TypeBlockBuilderPositionsAppender(Type type, BlockPositionEqual equalOperator, BlockBuilderStatus blockBuilderStatus, int expectedPositions)
+        {
+            super(type, equalOperator, blockBuilderStatus, expectedPositions);
+        }
+
+        @Override
+        protected void appendTo(IntArrayList positions, Block source, BlockBuilder blockBuilder)
+        {
+            int[] positionArray = positions.elements();
+            for (int i = 0; i < positions.size(); i++) {
+                type.appendTo(source, positionArray[i], blockBuilder);
+            }
+        }
+
+        @Override
+        protected void appendRle(int positionCount, Block source, int sourcePosition, BlockBuilder blockBuilder)
```

Comment:
I don't see a need for `appendRle` and `appendDictionary` methods. Instead, you should add `DictionaryAwarePositionsAppender` that would generically handle RLEs/dictionaries (based on instanceof)

Reply:
I added `appendRle` and `appendDictionary` because: 
a) in the case where the implementation is exactly the same as for `appendTo`, it creates a place for jit to optimize the code only for a single type without the overhead
b) there can be manually optimized implementations specifically for RLE and Dictionary. For RLE it's already optimized for multiple types.

How can `DictionaryAwarePositionsAppender` handle this without the overhead?

Reply:
> a) in the case where the implementation is exactly the same as for appendTo, it creates a place for jit to optimize the code only for a single type without the overhead

no need to specialize batch method call. per row is a different matter

> How can DictionaryAwarePositionsAppender handle this without the overhead?

```
class DictionaryAwarePositionsAppender implements BatchAppender {
  Block singleValue;
  int rowCount;
  BatchAppender delegate;

  protected void appendTo(IntArrayList positions, Block source) {
    if (source instanceof RunLengthEncodedBlock) {
      if (valuesEqual(source, singleValue)) {
        rowCount += positions.length();
      } else {
        Block flatBlock = // flatten singleValue into primitive block
        delegate.appendTo([0...rowCount], flatBlock);
        delegete.appendTo(positions, source);
      }
    } else {
      delegate.appendTo(positions, source);
    }
  }
}
```

but then it seems your methods make it cleaner:
```
class DictionaryAwarePositionsAppender implements BatchAppender {
  Block singleValue;
  int rowCount;
  BatchAppender delegate;

  protected void appendTo(IntArrayList positions, Block source) {
    if (source instanceof RunLengthEncodedBlock) {
      if (valuesEqual(source, singleValue)) {
        rowCount += positions.length();
      } else {
        delegate.appendRle(new RunLengthEncodedBlock(singleValue, rowCount);
        delegete.appendTo(positions, source);
      }
    } else {
      delegate.appendTo(positions, source);
    }
  }
}
```